src/server/transport.rs

```diff
@@ -106,100 +99,205 @@ impl RaftStoreRouter for ServerRaftStoreRouter {
+            raft_router: raft_router,
+            snapshot_status_sender: snapshot_status_sender,
+            resolving: Arc::new(RwLock::new(Default::default())),
+            resolver: Arc::new(Mutex::new(resolver)),
+        }
+    }
 
-        try!(self.ch.try_send(Msg::SendStore {
-            store_id: to_store_id,
-            msg: msg,
-        }));
-        Ok(())
+    fn send_store(&self, store_id: u64, msg: RaftMessage) {
+        // check the corresponding token for store.
+        let addr = self.raft_client.rl().addrs.get(&store_id).cloned();
+        if let Some(addr) = addr {
+            self.write_data(store_id, addr, msg);
+            return;
+        }
+
+        // No connection, try to resolve it.
+        if self.resolving.rl().contains(&store_id) {
+            RESOLVE_STORE_COUNTER.with_label_values(&["resolving"]).inc();
+            // If we are resolving the address, drop the message here.
+            debug!("store {} address is being resolved, drop msg {:?}",
+                   store_id,
+                   msg);
+            self.report_unreachable(msg);
+            return;
+        }
+
+        debug!("begin to resolve store {} address", store_id);
+        let label = if msg.get_message().has_snapshot() {
+            "snap"
+        } else {
+            "store"
+        };
+        RESOLVE_STORE_COUNTER.with_label_values(&[label]).inc();
+
+        self.resolving.wl().insert(store_id);
+        self.resolve(store_id, msg);
     }
-}
 
+    fn resolve(&self, store_id: u64, msg: RaftMessage) {
+        let trans = self.clone();
+        let cb = box move |addr| {
+            // clear resolving.
+            trans.resolving.wl().remove(&store_id);
+
+            if let Err(e) = addr {
+                RESOLVE_STORE_COUNTER.with_label_values(&["failed"]).inc();
+                debug!("resolve store {} address failed {:?}", store_id, e);
+                trans.report_unreachable(msg);
+                return;
+            }
+
+            RESOLVE_STORE_COUNTER.with_label_values(&["success"]).inc();
+            let addr = addr.unwrap();
+            info!("resolve store {} address ok, addr {}", store_id, addr);
+            trans.raft_client.wl().addrs.insert(store_id, addr);
+            trans.write_data(store_id, addr, msg);
+        };
+        if let Err(e) = self.resolver.lock().unwrap().resolve(store_id, cb) {
+            error!("try to resolve err {:?}", e);
+        }
+    }
 
-// MockRaftStoreRouter is used for passing compile.
-#[derive(Clone)]
-pub struct MockRaftStoreRouter;
+    fn write_data(&self, store_id: u64, addr: SocketAddr, msg: RaftMessage) {
+        if msg.get_message().has_snapshot() {
+            return self.send_snapshot_sock(addr, msg);
+        }
 
-impl RaftStoreRouter for MockRaftStoreRouter {
-    fn send(&self, _: StoreMsg) -> RaftStoreResult<()> {
-        unimplemented!();
+        if let Err(e) = self.raft_client.wl().send(store_id, addr, msg) {
```

Comment:
_send_ may need to change the inner state of _raft_client_ by creating new connection.